Fixes #6751 - ModelSerializer fields does not get updated correctly when signals are connected to some fields

Krystofee opened the pull request:
## Description

Many to many fields are updated separately after other serializer fields in `ModelSerializer`. Please see issue #6751 for more information.

refs #6751

Krystofee commented:
@tomchristie I know that this "seems like a pretty fundamental change", but since validated_data is instance of `dict`, the order of `validated_data.items()` did not matter. Everything I've done is that many-to-many fields are updated right after every other fields without changing any other functionality. I consider this as a good idea since updating m2m fields can trigger signals and cause side effects.

rpkilby commented on rest_framework/serializers.py:
```diff
@@ -973,12 +973,17 @@ def update(self, instance, validated_data):
-                field = getattr(instance, attr)
-                field.set(value)
-            else:
+            if attr not in info.relations or not info.relations[attr].to_many:
```
I would revert the condition, as the previous version is more readable to me.

Krystofee replied:
Good point, I've reverted the condition in the 064f1e9a146e323f6c6e7c252243ac4702827a38.

tomchristie commented:
This looks really good actually - nice one! 😀
We'll probably need to add a comment about *why* we set any m2m fields after everything else.
Ideally a test case that ensures we don't revert this behavior would be super helpful, but if it's *really* awkward then we could potentially pass on it.

Krystofee commented:
@tomchristie thx. I've added comment why this is desired behaviour in 4871372e2490cd042a9f1d7836bd191bab202e07. Also I've changed that instance is daved before setting m2m fields as @rpkilby suggested (https://github.com/encode/django-rest-framework/issues/6751#issuecomment-503674939). 

@tomchristie Though I'm not really sure where to include the test that will ensure this behaviour. I think I can add it when you tell me where.

rpkilby commented:
@Krystofee An appropriate place would be in [tests/test_model_serializer.py](https://github.com/encode/django-rest-framework/blob/master/tests/test_model_serializer.py), however I don't see an existing test case you can build off of.

Krystofee commented:
I've added the test for this behaviour. Note that the previous behaviour is dependent on the arbitrary order of the dict keys so it can pass even without that change. But I've ran the test against previous code and it failed so we can hope the order of the dict keys won't change across runs and it will not lead to any false positives. I did not figure out any better way to test it...

rpkilby commented:
I think this is sufficient. On Python 3.6 and later, keys _are_ ordered, so this test should consistently fail on modern versions without the fix. 

I think this is good to go. Thoughts @tomchristie?

tomchristie commented:
Yup. Thanks!
